components/num-util/src/fixed_decimal.rs

```diff
@@ -415,94 +420,123 @@ impl FromStr for FixedDecimal {
                         return Err(Error::Syntax);
                     }
                 }
-            } else if !c.is_digit(10) {
+            } else if c < b'0' || c > b'9' {
                 return Err(Error::Syntax);
             }
         }
-        // no_dot_str: shows the string when the dot is removed from it
-        let no_dot_str = no_sign_str.replace(".", "");
-        let no_dot_str_len = no_dot_str.chars().count();
+
+        // defining the output dec here and set its sign
+        let mut dec: FixedDecimal = 0.into();
+        dec.is_negative = is_negative;
+
+        // no_dot_str_len: shows length of the string after removing the dot
+        let mut no_dot_str_len = no_sign_str_len;
+        if has_dot {
+            no_dot_str_len -= 1;
+        }
 
         // left_zeros: number of zeros at the left of no_dot_str
         // right_zeros: number of zeros at the right of no_dot_str
-        // These are not necessarily leading or trailing zeros
+        // leftmost_digit: index of the first non-zero digit
+        // leftmost_digit: index of the last non-zero digit
+        // left zeros and right zeros are not necessarily leading or trailing zeros
         // Example:
-        //     input string     left_zeros      right_zeros
-        //     00123000             2                3
-        //     0.0123000            2               3
-        //     001.23000            2               3
-        //     001230.00            2               3
+        //     input string     left_zeros      right_zeros    leftmost_digit     rightmost_digit
+        //     00123000             2               3                 2                  4
+        //     0.0123000            2               3                 2                  4
```

Comment:
You are right about 'rightmost_digit', it is the index of the first element on the right side of 'digits' and should be 5 here. 'leftmost_digit' is in fact '2' here, because it represents the index of the first element of 'digits'. (Tested)